udao/optimization/soo/mogd.py

```diff
@@ -426,111 +455,102 @@ def _processed_single_start_opt(
 
-            # Update input_vars_subvector with constrained values
-            input_vars_subvector.data = th.clip(
-                input_vars_subvector.data,
-                # use .data to avoid gradient tracking during update
-                lower_input.features[0, grad_indices],
-                upper_input.features[0, grad_indices],
-            )
-            if i > best_iter + self.patience:
+            with th.no_grad():
+                # Update input_vars_subvector with constrained values
+                input_vars_subvector.data = th.clip(
+                    input_vars_subvector.data,
+                    # Use .data to avoid gradient tracking during update
+                    lower_input.features[0, grad_indices],
+                    upper_input.features[0, grad_indices],
+                )
+
+                if self.strict_rounding:
+                    # Round all integer variables at each iteration
+                    input_data.features[:, grad_indices] = input_vars_subvector.data
+                    feature_container = make_tabular_container(
+                        input_data.features.detach()
+                    )
+                    best_raw_df = problem.data_processor.inverse_transform(
+                        feature_container, "tabular_features"
+                    )
+                    numeric_values: Dict[str, np.ndarray] = {
+                        name: best_raw_df[[name]].values.round()[:, 0]
+                        if isinstance(variable, co.IntegerVariable)
+                        else best_raw_df[[name]].values[:, 0]
+                        for name, variable in problem.variables.items()
+                    }
+                    input_data_raw, _ = derive_processed_input(
+                        data_processor=problem.data_processor,
+                        input_parameters=problem.input_parameters or {},
+                        input_variables=numeric_values,
+                        device=self.device,
+                    )
+                    input_vars_subvector.data = input_data_raw.features[:, grad_indices]
+
+            if best_iter is not None and i > best_iter + self.patience:
                 break
             i += 1
-        if best_obj is not None and best_raw_vars is not None:
-            self._log_success(problem, i, best_obj, best_iter, best_raw_vars)
 
-            return best_obj, best_raw_vars, best_loss
+        if best_iter is not None:
+            assert best_obj is not None and best_feature_input is not None
+            with th.no_grad():
+                best_feature_input = cast(th.Tensor, best_feature_input)
+                feature_container = make_tabular_container(best_feature_input)
+                best_raw_df = problem.data_processor.inverse_transform(
+                    feature_container, "tabular_features"
+                )
+                if not self.strict_rounding:
+                    best_raw_vars: Dict[str, Any] = {
```

Comment:
In both processed and unprocessed case, rounding could be factored into a rounding function since if is repeated twice in the code.

Reply:
postpone this to a one-loop PR in the future